### PR TITLE
[CELEBORN-1118] Support memory cache for local disk files

### PR DESCRIPTION
**Pull request by oOKevinOo**
### What changes were proposed in this pull request?

Support for small file caching to improve shuffle performance,The architecture is as follows:
image

### Why are the changes needed?

Support for small file caching to improve shuffle performance

### Does this PR introduce _any_ user-facing change?

None

### How was this patch tested?
The test resource used is as follows:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/xxxxx/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/xxxxx/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

host memory | host cpu | disk | service heap memory | service off-heap memory | worker | master | spark total resource
-- | -- | -- | -- | -- | -- | -- | --
128G | 64 core | 11 * 7.1T | 12g | 40g | 5 | 3 | 25T/6400C

</body>

</html>

When turning off memory cache, test the performance as follows：
image
Duration of application execution：
image

The performance when I use memory for caching is as follows：
image
image
Duration of application execution：
image

The peak flow rate exceeds 20G, and the maximum time consumption is reduced to less than 2min,up to 4x performance improvement.

I think there are two aspects to improving performance:
1. Using memory reduces requests to obtain local disk data
2. Reduced file open consume and improved disk sequential read capability

### CODE REVIEW

File: common/src/main/java/org/apache/celeborn/common/util/MemCacheManager.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    cacheEnable = conf.workerMemoryCacheEnabled() && !conf.hasHDFSStorage();
+  }
+
+  public void putCache(String key, ByteBuf cache) {
+    int cacheSize = cache.readableBytes();
+    caches.put(key, cache);
+    currentCacheSize.getAndAdd(cacheSize);
+  }
+
+  public boolean canCache(int cacheSize) {
```
Comment (FMX) [inline common/src/main/java/org/apache/celeborn/common/util/MemCacheManager.java:46]: Here is a concurrent issue. Multiple threads might try to flush different file writers.  You should invoke `canCache` and `putCache` atomically.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionFilesSorter.java [hunk 2]
```diff
@@ -637,6 +639,12 @@ private void initializeFiles() throws IOException {
         hdfsSortedOutput =
             StorageManager.hadoopFs().create(new Path(sortedFilePath), true, 256 * 1024);
       } else {
+        if (memCacheManager.contains(originFilePath)) {
+          FileChannel channel = FileChannelUtils.createWritableFileChannel(originFilePath);
+          channel.write(memCacheManager.getCache(originFilePath).nioBuffer());
```
Comment (FMX) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionFilesSorter.java:644]: The shuffle file is in memory. I think you can sort the file in memory and return the sorted shuffle data.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java [hunk 3]
```diff
@@ -159,14 +162,24 @@ protected void flush(boolean finalFlush) throws IOException {
...
-          FlushTask task = null;
-          if (channel != null) {
-            task = new LocalFlushTask(flushBuffer, channel, notifier);
-          } else if (fileInfo.isHdfs()) {
-            task = new HdfsFlushTask(flushBuffer, fileInfo.getHdfsPath(), notifier);
+          if (finalFlush
+              && channel != null
+              && fileInfo.getFileLength() == 0
+              && cacheManager.canCache(numBytes)) {
+            cacheManager.putCache(fileInfo.getFilePath(), flushBuffer.copy());
```
Comment (FMX) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java:169]: Actually, this copy is unnecessary. You can just use this buffer and return it when cache is released.

File: common/src/main/java/org/apache/celeborn/common/meta/MemoryBuffers.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.celeborn.common.network.buffer.ManagedBuffer;
+import org.apache.celeborn.common.network.buffer.NettyManagedBuffer;
+import org.apache.celeborn.common.network.util.TransportConf;
+import org.apache.celeborn.common.util.MemCacheManager;
+
+public class MemoryBuffers implements Buffers {
```
Comment (FMX) [inline common/src/main/java/org/apache/celeborn/common/meta/MemoryBuffers.java:31]: This class can be merged into FileManagedBuffers. These codes look quite similar.

File: common/src/main/java/org/apache/celeborn/common/meta/Buffers.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.common.meta;
+
+import org.apache.celeborn.common.network.buffer.ManagedBuffer;
+
+public interface Buffers {
```
Comment (FMX) [inline common/src/main/java/org/apache/celeborn/common/meta/Buffers.java:22]: Unnecessary

File: common/src/main/java/org/apache/celeborn/common/meta/FileManagedBuffers.java [hunk 6]
```diff
@@ -24,7 +24,8 @@
 import org.apache.celeborn.common.network.buffer.ManagedBuffer;
 import org.apache.celeborn.common.network.util.TransportConf;
 
-public class FileManagedBuffers {
+public class FileManagedBuffers implements Buffers {
```
Comment (FMX) [inline common/src/main/java/org/apache/celeborn/common/meta/FileManagedBuffers.java:27]: Making this class able to read shuffle files and shuffle memory cache will make code more readable.

File: common/src/main/java/org/apache/celeborn/common/util/MemCacheManager.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+import io.netty.buffer.ByteBuf;
+
+import org.apache.celeborn.common.CelebornConf;
+
+public class MemCacheManager {
```
Comment (FMX) [inline common/src/main/java/org/apache/celeborn/common/util/MemCacheManager.java:28]: This class can be merged into memory manager.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionFilesSorter.java (hunk 2 above)
Comment (waitinfuture) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionFilesSorter.java:644]: IMO supporting sort in memory cache is out of this PR's scope, I think this is ok to fallback to disk-based when sort happens.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java (hunk 3 above)
Comment (waitinfuture) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java:169]: copy() here is to simplify release logic

**Pull request closed without merging**